File: mobile/lib/features/jobs/job_status.dart
import 'package:flutter/material.dart';

import '../../core/theme.dart';

/// Metadata status job v2 (PRD §6) dan slot foto (BR-06) — satu sumber
/// kebenaran untuk label, warna, urutan tahap, dan slot wajib.
enum JobStatus {
  menungguPickup('menunggu_pickup'), // nilai lama, diperlakukan sebagai ditugaskan
  ditugaskan('ditugaskan'),
  diterima('diterima'),
  loading('loading'),
  dalamPerjalanan('dalam_perjalanan'),
  unloading('unloading'),
  serahTerimaPool('serah_terima_pool'),
  menungguValidasi('menunggu_validasi'),
  selesai('selesai'),
  cancelled('cancelled');

  const JobStatus(this.value);
  final String value;

  static JobStatus parse(String raw) => JobStatus.values.firstWhere(
        (s) => s.value == raw,
        orElse: () => JobStatus.ditugaskan,
      );

  /// Status lama `menunggu_pickup` dinormalkan ke `ditugaskan`.
  JobStatus get normalized => this == JobStatus.menungguPickup ? JobStatus.ditugaskan : this;

  bool get isClosed => this == JobStatus.selesai || this == JobStatus.cancelled;
  bool get isActive => !isClosed;

  String get label => switch (normalized) {
        JobStatus.ditugaskan => 'Ditugaskan',
        JobStatus.diterima => 'Diterima',
        JobStatus.loading => 'Loading',
        JobStatus.dalamPerjalanan => 'Dalam perjalanan',
        JobStatus.unloading => 'Unloading',
        JobStatus.serahTerimaPool => 'Serah terima pool',
        JobStatus.menungguValidasi => 'Menunggu validasi',
        JobStatus.selesai => 'Selesai',
        JobStatus.cancelled => 'Dibatalkan',
        JobStatus.menungguPickup => 'Ditugaskan',
      };

  Color get color => switch (normalized) {
        JobStatus.ditugaskan => MasColors.warning,
        JobStatus.diterima => MasColors.info,
        JobStatus.loading || JobStatus.unloading => MasColors.info,
        JobStatus.dalamPerjalanan => MasColors.brandDark,
        JobStatus.serahTerimaPool => MasColors.info,
        JobStatus.menungguValidasi => MasColors.warning,
        JobStatus.selesai => MasColors.brand,
        JobStatus.cancelled => MasColors.danger,
        JobStatus.menungguPickup => MasColors.warning,
      };

  Color get background => switch (normalized) {
        JobStatus.ditugaskan || JobStatus.menungguValidasi => MasColors.warningBg,
        JobStatus.selesai || JobStatus.dalamPerjalanan => MasColors.brandLight,
        JobStatus.cancelled => MasColors.dangerBg,
        _ => MasColors.infoBg,
      };

  /// Status berikutnya yang boleh diminta driver (FR-MOBILE-06).
  JobStatus? get nextDriverStatus => switch (normalized) {
        JobStatus.diterima => JobStatus.loading,
        JobStatus.loading => JobStatus.dalamPerjalanan,
        JobStatus.dalamPerjalanan => JobStatus.unloading,
        JobStatus.unloading => JobStatus.serahTerimaPool,
        JobStatus.serahTerimaPool => JobStatus.menungguValidasi,
        _ => null,
      };

  /// Teks tombol lanjut per status.
  String? get advanceLabel => switch (normalized) {
        JobStatus.diterima => 'Tiba di lokasi muat — Mulai Loading',
        JobStatus.loading => 'Muat selesai — Berangkat',
        JobStatus.dalamPerjalanan => 'Tiba di tujuan — Mulai Bongkar',
        JobStatus.unloading => 'Bongkar selesai — Kembali ke Pool',
        JobStatus.serahTerimaPool => 'Selesaikan Orderan',
        _ => null,
      };

  /// Tahap foto yang sedang dibuka pada status ini (foto hanya boleh diambil
  /// saat status sudah masuk tahapnya).
  PhotoStage? get openPhotoStage => switch (normalized) {
        JobStatus.loading => PhotoStage.loading,
        JobStatus.unloading => PhotoStage.unloading,
        JobStatus.serahTerimaPool => PhotoStage.serahTerima,
        _ => null,
      };
}

/// Urutan langkah yang ditampilkan di stepper driver.
const driverSteps = <JobStatus>[
  JobStatus.ditugaskan,
  JobStatus.diterima,
  JobStatus.loading,
  JobStatus.dalamPerjalanan,
  JobStatus.unloading,
  JobStatus.serahTerimaPool,
  JobStatus.menungguValidasi,
  JobStatus.selesai,
];

enum PhotoStage {
  loading('loading', 'Loading (muat)'),
  unloading('unloading', 'Unloading (bongkar)'),
  serahTerima('serah_terima', 'Serah terima di pool');

  const PhotoStage(this.value, this.label);
  final String value;
  final String label;

  static PhotoStage? tryParse(String? raw) {
    for (final s in PhotoStage.values) {
      if (s.value == raw) return s;
    }
    return null;
  }

  /// Slot yang wajib terisi (BR-06).
  List<PhotoSlot> get requiredSlots => switch (this) {
        PhotoStage.loading || PhotoStage.unloading => const [
            PhotoSlot.depan,
            PhotoSlot.belakang,
            PhotoSlot.kanan,
            PhotoSlot.kiri,
            PhotoSlot.suratTimbang,
          ],
        PhotoStage.serahTerima => const [PhotoSlot.serahTerima],
      };
}

enum PhotoSlot {
  depan('depan', 'Foto sisi depan kendaraan'),
  belakang('belakang', 'Foto sisi belakang kendaraan'),
  kanan('kanan', 'Foto sisi kanan kendaraan'),
  kiri('kiri', 'Foto sisi kiri kendaraan'),
  suratTimbang('surat_timbang', 'Foto surat timbang'),
  serahTerima('serah_terima', 'Foto serah terima dokumen');

  const PhotoSlot(this.value, this.label);
  final String value;
  final String label;

  static PhotoSlot? tryParse(String? raw) {
    for (final s in PhotoSlot.values) {
      if (s.value == raw) return s;
    }
    return null;
  }

  /// Surat timbang harus lebih tajam karena berisi teks (FR-PHOTO-05).
  bool get isDocument => this == PhotoSlot.suratTimbang;
}

/// Teks kunci yang ditampilkan ke driver (FR-UJ-07).
const lockMessageUangJalan = 'Menunggu admin mengunggah bukti transfer uang jalan.';

/// Teks alert pengajuan uang jalan (PRD §6.2 — persis).
const ajukanUangJalanAlert =
    'Apakah anda yakin ingin mengajukan uang jalan? Anda baru bisa melanjutkan perjalanan '
    'setelah admin kasir mengupload bukti transfer uang jalan.';

File: frontend/src/lib/job-status.ts
/**
 * Metadata status job v2 (PRD-Alur-Kerja-Job-v2 §6) dan slot foto (BR-06).
 * Satu tempat supaya label, warna, dan urutan tahap tidak tersebar.
 */

import type { JobStatus, PhotoSlot, PhotoStage } from "@/types";

export const JOB_STATUS_LABEL: Record<JobStatus, string> = {
  menunggu_pickup: "Ditugaskan",
  ditugaskan: "Ditugaskan",
  diterima: "Diterima driver",
  loading: "Loading",
  dalam_perjalanan: "Dalam perjalanan",
  unloading: "Unloading",
  serah_terima_pool: "Serah terima pool",
  menunggu_validasi: "Menunggu validasi",
  selesai: "Selesai",
  cancelled: "Dibatalkan"
};

/** Kelas badge (lihat globals.css `.badge-*`). */
export const JOB_STATUS_BADGE: Record<JobStatus, string> = {
  menunggu_pickup: "badge-pickup",
  ditugaskan: "badge-pickup",
  diterima: "badge-pickup",
  loading: "badge-loading",
  dalam_perjalanan: "badge-perjalanan",
  unloading: "badge-unloading",
  serah_terima_pool: "badge-unloading",
  menunggu_validasi: "badge-perbaikan",
  selesai: "badge-selesai",
  cancelled: "badge-cancelled"
};

export const JOB_STATUS_COLOR: Record<JobStatus, { bg: string; fg: string }> = {
  menunggu_pickup: { bg: "var(--status-pickup-bg)", fg: "var(--status-pickup-text)" },
  ditugaskan: { bg: "var(--status-pickup-bg)", fg: "var(--status-pickup-text)" },
  diterima: { bg: "var(--status-pickup-bg)", fg: "var(--status-pickup-text)" },
  loading: { bg: "#fff4e0", fg: "#8a5a00" },
  dalam_perjalanan: { bg: "var(--status-perjalanan-bg)", fg: "var(--status-perjalanan-text)" },
  unloading: { bg: "#efeafe", fg: "#4a2bb0" },
  serah_terima_pool: { bg: "#efeafe", fg: "#4a2bb0" },
  menunggu_validasi: { bg: "var(--status-perbaikan-bg)", fg: "var(--status-perbaikan-text)" },
  selesai: { bg: "var(--status-selesai-bg)", fg: "var(--status-selesai-text)" },
  cancelled: { bg: "var(--status-cancelled-bg)", fg: "var(--status-cancelled-text)" }
};

/** Warna solid untuk papan jadwal. */
export const JOB_STATUS_SOLID: Record<JobStatus, string> = {
  menunggu_pickup: "#8A8A85",
  ditugaskan: "#8A8A85",
  diterima: "#1f4fa8",
  loading: "#c97900",
  dalam_perjalanan: "#1c9600",
  unloading: "#4a2bb0",
  serah_terima_pool: "#4a2bb0",
  menunggu_validasi: "#854f0b",
  selesai: "#145b00",
  cancelled: "#c93030"
};

/** Status yang masih menahan unit & driver (BR-01). */
export const ACTIVE_JOB_STATUSES: JobStatus[] = [
  "menunggu_pickup",
  "ditugaskan",
  "diterima",
  "loading",
  "dalam_perjalanan",
  "unloading",
  "serah_terima_pool",
  "menunggu_validasi"
];

export function isActiveStatus(status: JobStatus): boolean {
  return ACTIVE_JOB_STATUSES.includes(status);
}

/** Status yang bisa dilalui driver berikutnya (satu langkah maju). */
export const NEXT_DRIVER_STATUS: Partial<Record<JobStatus, JobStatus>> = {
  diterima: "loading",
  loading: "dalam_perjalanan",
  dalam_perjalanan: "unloading",
  unloading: "serah_terima_pool",
  serah_terima_pool: "menunggu_validasi"
};

/** Tahap yang ditampilkan ke pelanggan — proses internal (pool/validasi) disembunyikan. */
export const CUSTOMER_STEPS: Array<{ key: JobStatus; label: string }> = [
  { key: "ditugaskan", label: "Dijadwalkan" },
  { key: "loading", label: "Loading" },
  { key: "dalam_perjalanan", label: "Dalam perjalanan" },
  { key: "unloading", label: "Unloading" },
  { key: "selesai", label: "Selesai" }
];

/** Petakan status internal ke tahap pelanggan. */
export function customerStep(status: JobStatus): JobStatus {
  switch (status) {
    case "menunggu_pickup":
    case "diterima":
      return "ditugaskan";
    case "serah_terima_pool":
    case "menunggu_validasi":
      return "unloading";
    default:
      return status;
  }
}

// ── Foto per slot (BR-06) ────────────────────────────────────────────────────

export const SLOT_LABEL: Record<PhotoSlot, string> = {
  depan: "Foto sisi depan kendaraan",
  belakang: "Foto sisi belakang kendaraan",
  kanan: "Foto sisi kanan kendaraan",
  kiri: "Foto sisi kiri kendaraan",
  surat_jalan: "Foto surat jalan",
  serah_terima: "Foto serah terima dokumen"
};

export const STAGE_LABEL: Record<PhotoStage, string> = {
  loading: "Loading (muat)",
  unloading: "Unloading (bongkar)",
  serah_terima: "Serah terima di pool"
};

export const REQUIRED_SLOTS: Record<PhotoStage, PhotoSlot[]> = {
  loading: ["depan", "belakang", "kanan", "kiri", "surat_jalan"],
  unloading: ["depan", "belakang", "kanan", "kiri", "surat_jalan"],
  serah_terima: ["serah_terima"]
};

/** Tahap foto yang relevan untuk sebuah status (untuk tampilan driver). */
export function stageForStatus(status: JobStatus): PhotoStage | null {
  switch (status) {
    case "diterima":
    case "loading":
      return "loading";
    case "dalam_perjalanan":
    case "unloading":
      return "unloading";
    case "serah_terima_pool":
      return "serah_terima";
    default:
      return null;
  }
}
